### ISSUE TITLE
Using numba for optimizing the estimator functions code

### ISSUE THREAD

**Issue by mralbu** [LSchueler/GSTools#6]
I had some issues and did not manage to compile the cython code on Windows. 
I've converted the estimator.pyx cython code to numba. The numba code seems to run reasonably fast without major optimizations, but I haven't thoroughly compared execution times. 

[Numba branch](https://github.com/mralbu/GSTools/tree/Numba)

**Comment by LSchueler** [comment]
Thanks a lot for your contribution!

Could you describe your problem with Cython under Windows a bit more in detail? - It would be nice to figure out what the problem is.

We still haven't finally decided if we want to use Cython for future functionality. If not, it would make sense to completely abandon Cython for easier deployment.
Your issue initiated @MuellerSeb to see if he can come up with a vectorised algorithm in pure numpy, without making it totally unreadable (like my last try). If that works out, we could use that as the default and use cython / numba in case the memory demand is too large.

If we use numba, it would be nice if you would ad the signatures of the functions.

**Comment by mralbu** [comment]
Actually, I think the problem I faced was with the cython and gcc/Mingw installation on this particular machine and not with GSTools itself. 
A pure numpy or numba implementation would ease such problems, and may simplify code.
Taking a quick look at the functions signatures, I see that some of them do not use all of their parameters (for instance, structured_2d(x, y, f) does not depend on x, y). I will try to understand better how they are called, maybe they can be implemented using np.diff functions.

By the way, would like to point out that I really appreciate that you've open-sourced GSTools. I am using it to generate synthetic datasets for fluid flow simulations. Thank you!

**Comment by MuellerSeb** [comment]
I got GSTools installed under Windows. 
I used WinPython and installed Visual Studio Build Tools as described in this Post: https://stackoverflow.com/a/50210015.

I would think of that as a standard situation to get Cython to work under Windows and I think, that should be it. I like the Idea of Numba, but I would like to keep Cython as the standard situation. And so we avoid getting another dependency. But I really would like to have a look at your log-file/error-message, to see what went wrong with Cython in your case!

Maybe we could provide pre-compiled libs in the future, so Cython doesn't actually need to be run under windows.

**Comment by mralbu** [comment]
Pre-built windows libs would be great! I don't have admin rights on this specific Windows machine, and because of that I was not able to install VS Build Tools. I managed to install mingw / gcc using anaconda, but it may not have been properly set up, and it did not manage to compile the package's cython code. Pre-compiled libs would be a great help!

**Comment by MuellerSeb** [comment]
OK, so this will be the way to go! 
@LSchueler : Do you mind to care about these pre-builts?
@mralbu : Thank you for the interesting input!

**Comment by LSchueler** [comment]
@MuellerSeb yes, I opened a new issue #7 and assigned myself to it. But I will need some time for that. The topic seems pretty complicated and it's new to me. But I agree, it's important.

@mralbu I updated the master branch today and included a check if the Cython stuff was successfully compiled. If not, the pure Python implementation from @MuellerSeb will be used. Although very elegant, the problem with that version is the huge memory usage.
But now that I set that up, we can easily incorporate your Numba implementation and also ignore it, in case Numba isn't working on a machine. Would you like to contribute your code?

By the way, I have an old version of a field generator lying around, which generates random velocity fields in a first order approximation. If you are interested in it, I could bring it up to date and upload it to this repository.

**Comment by MuellerSeb** [comment]
The numpy implementation could also just use chunks like the RandMeth generator does with the modes ([see here](https://github.com/LSchueler/GSTools/blob/1814474ddf5eb138e973941962cbd00ff498b59d/gstools/field/srf.py#L439)), since the variogram is a simple sum for each bin. Maybe this could solve some problems.. I'll have a look at that later.

**Comment by MuellerSeb** [comment]
#7 is now closed. Prebuild wheels are now provided for py27, py35, py36 on linux, macOS, and Windows for 32 and 64 bit each. Hope this helps with your case!

**Comment by MuellerSeb** [comment]
@mralbu : The 1.0 Version is now released. Just try it if you want!

**Comment by mralbu** [comment]
Great, will try it!! :smiley:
